pkg/util/http.go

```diff
@@ -195,60 +200,72 @@ func (e MsgSizeTooLargeErr) Is(err error) bool {
 	return ok1 || ok2
 }
 
-func decompressRequest(dst []byte, reader io.Reader, expectedSize, maxSize int, compression CompressionType, sp opentracing.Span) (body []byte, err error) {
-	defer func() {
-		if err != nil && len(body) > maxSize {
-			err = MsgSizeTooLargeErr{Actual: len(body), Limit: maxSize}
-		}
-	}()
+func decompressRequest(dst []byte, reader io.Reader, expectedSize, maxSize int, compression CompressionType, sp opentracing.Span) ([]byte, error) {
 	if expectedSize > maxSize {
 		return nil, MsgSizeTooLargeErr{Actual: expectedSize, Limit: maxSize}
 	}
-	buffer, ok := tryBufferFromReader(reader)
-	if ok {
-		body, err = decompressFromBuffer(dst, buffer, maxSize, compression, sp)
-		return
-	}
-	body, err = decompressFromReader(dst, reader, expectedSize, maxSize, compression, sp)
-	return
-}
 
-func decompressFromReader(dst []byte, reader io.Reader, expectedSize, maxSize int, compression CompressionType, sp opentracing.Span) ([]byte, error) {
-	var (
-		buf  bytes.Buffer
-		body []byte
-		err  error
-	)
-	if expectedSize > 0 {
-		buf.Grow(expectedSize + bytes.MinRead) // extra space guarantees no reallocation
-	}
-	// Read from LimitReader with limit max+1. So if the underlying
-	// reader is over limit, the result will be bigger than max.
-	reader = io.LimitReader(reader, int64(maxSize)+1)
 	switch compression {
-	case NoCompression:
-		_, err = buf.ReadFrom(reader)
-		body = buf.Bytes()
-	case RawSnappy:
-		_, err = buf.ReadFrom(reader)
+	case NoCompression, RawSnappy:
+		buf, ok := tryBufferFromReader(reader)
+		if ok {
+			return decompressFromBuffer(dst, buf, maxSize, compression, sp)
+		}
+
+		buf = &bytes.Buffer{}
+		if expectedSize > 0 {
+			buf.Grow(expectedSize + bytes.MinRead) // extra space guarantees no reallocation
+		}
+
+		if _, err := buf.ReadFrom(reader); err != nil {
+			// TODO: Translate error
+			return nil, errors.Wrap(err, "read write request")
+		}
+
+		return decompressFromBuffer(dst, buf, maxSize, compression, sp)
+	case Gzip:
+		if sp != nil {
+			sp.LogFields(otlog.Event("util.ParseProtoReader[decompress]"),
+				otlog.Int("size", expectedSize))
+		}
+
+		var err error
+		reader, err = gzip.NewReader(reader)
 		if err != nil {
-			return nil, err
+			return nil, errors.Wrap(err, "create gzip reader")
 		}
-		body, err = decompressFromBuffer(dst, &buf, maxSize, RawSnappy, sp)
+
+		// Limit at maxSize+1 so we can tell when the size is exceeded
+		reader = io.LimitReader(reader, int64(maxSize)+1)
```

Comment:
This limiting is on the _decompressed_ output though, whereas the limiting in `ParseProtoReader` is on the compressed output. The limiting here corresponds to the `if size > maxSize` check in `decompressFromBuffer`.

Reply:
@fayzal-g after reviewing how it was done before, I removed the pre-Gzip decompression limiting, since the main branch OTLP handler doesn't do this. As the PR is just supposed to refactor, it's pointless to change this behaviour regardless of whether it's a good idea or not (limiting before Gzip decompressing).